Support maximum of 64kB page on aarch64

yuyichao opened the pull request:
Build the preload library to be aligned to 64kB following the default of binutils.

I've never used capnp before and I'm not sure yet how to add and use a new header field.....

If I'm doing things correctly, this should not affect x86...

Fix #3143

yuyichao commented:
* I've cleaned up the commit a bit, now there are a few relatively standalone commits here that should hopefully help with reviewing. Most of the commits are refactoring that doesn't make a difference on x86.
* Some of the commits can be its own PR, one of which is https://github.com/rr-debugger/rr/pull/3154. Since I think this is fully ready now I'm not going to split any out ATM into their own PR.
* No separate linker script anymore. Use configure file instead.
* It should be relatively easy to change the build time page size in the future, as shown in the last commit.
* I've added two fields in the header for the runtime page size and the build time page size as suggested by @rocallahan

rocallahan commented on src/preload/syscall_hook.S:
```diff
@@ -1,4 +1,8 @@
+        .set preload_thread_locals,0x70010000
+#else
         .set preload_thread_locals,0x70001000
+#endif
```
Why doesn't this use `0x70000000 + RR_PAGE_SIZE`?

yuyichao replied:
This one was actually just because I was too lazy to check if the macro was defined by this point ;-p...

rocallahan commented on src/test/mmap_replace_most_mappings.c:
```diff
@@ -5,7 +5,11 @@
+#define RR_THREAD_LOCALS_PAGE_ADDR 0x70010000
+#else
 #define RR_THREAD_LOCALS_PAGE_ADDR 0x70001000
+#endif
```
Likewise this could be `RR_PAGE_ADDR + page_size()`?

yuyichao replied:
This is in the test so it's somewhat different. It would have to be `rr_page_size()` as well.

rocallahan reviewed (changes requested):
I think you mean librrpreload.so, not librrpage.so?

I think we should move `rr_page_size()` to to `util.h`/`util.cc` as a global function that's a peer of `page_size()`. I think we should also change its name ... how about `rr_max_supported_page_size()` since it's the the maximum page size supported by this build of rr (which is baked into librrpreload.so layout)?

yuyichao commented:
I'm fine with renaming. I pick the current name mainly to emphasize that it's the size of the block in the preload library, since that's actually what most of the code uses it for (they use that size with a multiplication factor to compute an offset/address). It's somewhat less obvious what `address + n * maximum_support_page_size` means but as I said, I don't feel too strong about it either...

yuyichao commented:
I also realized that I left one instance of RR_PAGE_SIZE unchanged so I'll need to fix that...

I also did only minimum amount of refactoring but I could refactor it more to make most things that need a page size to get them from fewer places. I wasn't sure how much to abstract out/assume but I guess I'll just assume that we'll have an architecture dependent maximum page size to support and refactor everything around that. If there's an architecture that doesn't have a upper limit for page size it might be a problem but I think the system linker would also have to make such a decision anyway.....

rocallahan commented:
OK how about calling it `preload_library_page_size()` and document that it's the maximum supported page size for this architecture/rr build?

yuyichao commented:
Is there any specific reason to have both a function version `rr_page_size()` and a macro/constant version `RR_PAGE_SIZE` of seemingly the same thing? Should they be merged/both renamed?

rocallahan commented:
You're right. We should just have a `PRELOAD_LIBRARY_PAGE_SIZE` constant and use that everywhere.

yuyichao commented:
OK, so there's no need to make it `arch` dependent at runtime? I guess we can worry about that if/when there's support for cross architecture record/replay?...

rocallahan commented:
This has to be a build-time property because it determines the layout of the preload library.

yuyichao commented:
But in principle it only need to be a build time property `librrpage.so` (pretty sure this is the one I'm talking about, not librrpreload.so), the rr binary could accept multiple sizes and pick them at runtime.....

Anyway, constant it is for now...

rocallahan commented:
Right. I want it in the trace header so that later we can change the value for a given architecture and extend rr to support replaying "old" traces with a different value, so it would have to be a function outside the preload library. But we don't need to do that now.

yuyichao commented:
I updated with the new name and replaced all use of rr_page_size with the constant. As shown in the last commit, there are still a few (5) places where this value is defined. This is because, AFAICT, these do not include a shared header.

The simplest method to avoid this is to just add a `add_definitions(-DPRELOAD_LIBRARY_PAGE_SIZE=${PRELOAD_LIBRARY_PAGE_SIZE})`. Maybe even an `PRELOAD_LIBRARY_PAGE_SHIFT` so that `rr_page.S` doesn't have to hard code another value that may get out of sync. However, this would mean that the header would be less standalone which is a minor concern since preload_interface.h is installed. If someone is relying on preload_interface.h being usable (granted the renaming would have broke the API) an alternative is to use configure_file to create a header with this value. If that's not a concern I can just go with `add_definitions`.

yuyichao commented:
(The new field in the trace file header is also updated)

rocallahan commented:
I think this is good enough for now.

rocallahan commented:
Actually I added a followup patch 9d623e8a92498d3f6e79eb3afb73a5ac6af3f5a3 to set the default page size to its correct value, and made the page sizes UInt32.
